### PR TITLE
feat(model): confidence intervals

### PR DESCRIPTION
**Pull request by felipehuerta17**
Implemented random bootstrap and an example in the Jupyter notebook.

I created a new python module, post_regression.py, to encapsulate all the post-regression analysis such as confidence intervals, sensitivity analysis, among others.

Function usage:
```
from post_regression import ci_bootstrap
ci, par_list = ci_bootstrap(model, t_obs, n_I_obs, population)
```

Scalability: if the fit function is updated to update a user defined set of parameters, this functions should be slightly updated to reflect these changes.

Skew towards lower r0 values: This is expected as r0 has not been constant owing to recent lockdowns and social distancing measures.

### CODE REVIEW

File: post_regression.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+import model
+import numpy as np
+from sklearn.utils import resample
+
+def ci_bootstrap(model, t_obs, n_I_obs, population, alpha=0.95, n_iter=1000, r0_ci = True):
```
Comment (jia200x) [inline post_regression.py:5]: Should we move this to a [Mixin](https://easyaspython.com/mixins-for-fun-and-profit-cb9962760556) now or in a follow up?

Comment (jia200x) [reply post_regression.py:5]: This function and the one below have the same signature. I suggest to add a class member that allows to choose between different methods.
E.g `my_sir.ci(method="bootstrap", t_obs, ...)

File: post_regression.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    common use of this list is to plot histograms to visualize and
+    try to infer the probability density function of the parameters.
+    """
+
+    p0 = model.p 
+    w0 = model.w0
+
+    # Consider at least the three first datapoints
+    p_list = []
+    MSE_list = [] # List of mean squared errors of the prediction for the time t+1
```
Comment (jia200x) [inline post_regression.py:121]: please, rename these functions to match snake_case (lowercase and `_`)

File: post_regression.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        # Store the rolling parameters
+        p_list.append(model.p)
+        # Predict for the i+1 period
+        sol = model.solve(t_obs[i+lag], numpoints = t_obs[i+lag]+1)
+        # Calculate mean squared errors
+        MSE = np.sqrt( (population*sol[i-1+lag,2] - n_I_obs[i-1+lag])**2 )
+        MSE_list.append(MSE)
+
+    p_list = np.array(p_list)
+    MSE_list = np.array(MSE_list)
```
Comment (jia200x) [inline post_regression.py:134]: Same here. I suggest using `pylint` to check these errors (e.g `pylint model.py`)

**Comment by jia200x** [comment]
BTW, good practice of adding multi line commit messages!

File: post_regression.py (hunk 1 above)
Comment (felipehuerta17) [reply post_regression.py]: I agree. I think it is a good idea but for now they have different inputs and outputs. I updated that. Let's think on a better structure for the next PR

File: post_regression.py (hunk 2 above)
Comment (felipehuerta17) [reply post_regression.py]: Done!

**Comment by jia200x** [comment]
please rebase with the latest master

**Comment by jia200x** [comment]
Besides the rebase, I add some friendly reminders to help maintaining the code:

1. Please avoid rebasing and force pushing if there are pending issues (not marked as resolved). Otherwise the comments are lost and the reviewer needs to perform the whole review again.
An alternative is the usage of "fixups". Whenever there's a change to be introduced, it should be commited as a "fixup" commit. It's very common to commit with message "!fixup <commit_msg>". This can be done automatically using the --fixup option of git commit:
```
git commit --fixup <hash_of_the_commit_to_fix_up>
```

Then at a later step these fixups can be squashed to the original commit. If the `--fixup` feature is used, the squash can be done automatically with the `--autosquash` command of `git rebase`. E.g
```
git rebase -i <HASH_OF_THE_ROOT_OF_THE_BRANCH> --autosquash
```
The hash of the root of the branch is the commit where the branch starts (usually the one that appears in the master branch)

2. Although we are using the "squash and merge" feature of GitHub, PRs are easier to review if they have a few "semantics" commits. There should be one commit per "semantical" change rather than one per concept.
For instance, if a "pylint" rule is added and a new function with > 600 lines encapsulates a concept, then there should be one "chore" commit and one "feat".
If a change affects and existing component, it's good to mention that in a separate commit.
If there are changes to be done, please considering using fixups (see 1.)

3. It's a good practice to post WIP PRs when the PR is not ready (or require an unexpected change). One alternative is to write [WIP] in front of the description or add a WIP label to GitHub. So, reviewers can focus in other PRs while the original PR is getting in shape.

Let me know if there are any questions, as these procedures might be hard to understand at first sight.

**Comment by felipehuerta17** [comment]
rebased to master

**Comment by felipehuerta17** [comment]
I significantly messed up the version control history by committing in the middle of a rebase. This PR was closed but the latest changes are reflected on pr/confidence_intervals_v2
